tee-worker/omni-executor/binance-api/src/lib.rs

```diff
@@ -0,0 +1,161 @@
+	pub fn sign_request(&self, query_string: &str) -> String {
+		let mut mac = Hmac::<Sha256>::new_from_slice(self.api_secret.as_bytes())
+			.expect("HMAC can take key of any size");
+		mac.update(query_string.as_bytes());
+		let result = mac.finalize();
+		let code_bytes = result.into_bytes();
+		hex::encode(code_bytes)
+	}
+
+	/// Helper for making public API requests
+	pub async fn make_public_get_request<T>(
+		&self,
+		endpoint: &str,
+		parameters: Option<HashMap<String, String>>,
+	) -> Result<T, Error>
+	where
+		T: serde::de::DeserializeOwned,
+	{
+		let mut url = self.base_url.join(endpoint).unwrap();
+		if let Some(p) = parameters {
+			url.query_pairs_mut().extend_pairs(p.iter());
+		}
+
+		let response = self.client.get(url.as_str()).send().await.map_err(|e| {
+			error!("API request failed: {}", e);
+			Error::RequestFailed
+		})?;
+
+		response.json::<T>().await.map_err(|e| {
+			error!("Error parsing response: {}", e);
+			Error::ParseResponseFailed
+		})
+	}
+
+	/// Helper for making signed API requests
+	pub async fn make_signed_request<T>(
+		&self,
+		endpoint: &str,
+		method: Method,
+		parameters: Option<HashMap<String, String>>,
+		recv_window: Option<u32>,
+	) -> Result<T, Error>
+	where
+		T: serde::de::DeserializeOwned,
+	{
+		let mut url = self.base_url.join(endpoint).unwrap();
+		let mut params = HashMap::new();
+
+		let timestamp = SystemTime::now()
+			.duration_since(UNIX_EPOCH)
+			.expect("Time went backwards")
+			.as_millis() as u64;
+		params.insert("timestamp".to_string(), timestamp.to_string());
+
+		if let Some(window) = recv_window {
+			if window > MAX_RECV_WINDOW {
+				return Err(Error::MaxRecvWindowExceeded);
+			}
+			params.insert("recvWindow".to_string(), window.to_string());
+		}
+
+		if let Some(p) = parameters {
+			params.extend(p);
+		}
+
+		let signature = {
+			let mut tmp_url = url.clone();
+			tmp_url.query_pairs_mut().extend_pairs(params.iter());
+			let query_string = tmp_url.query().unwrap().to_string();
+			self.sign_request(&query_string)
+		};
+
+		let request = match method {
+			Method::GET => {
+				url.query_pairs_mut().extend_pairs(params.iter());
+				url.query_pairs_mut().append_pair("signature", &signature);
+				self.client.request(method, url.as_str()).header("X-MBX-APIKEY", &self.api_key)
+			},
+			Method::POST => {
+				params.insert("signature".to_string(), signature);
+				self.client
+					.request(method, url.as_str())
+					.header("X-MBX-APIKEY", &self.api_key)
+					.form(&params)
+			},
+			_ => {
+				return Err(Error::MethodNotSupported);
+			},
+		};
+
+		let response = request.send().await.map_err(|e| {
+			error!("API request failed: {}", e);
+			Error::RequestFailed
+		})?;
+
+		response.json::<T>().await.map_err(|e| {
+			error!("Error parsing response: {}", e);
+			Error::ParseResponseFailed
+		})
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+
+	#[test]
+	fn test_sign_request() {
+		let api_key =
```

Comment:
How are these api keys and secrets generated? Is it some known secrets (for tests) or did you generate them locally?

Reply:
These are known credentials, used in binance api documentation as an example to verify that the signature is generated correctly